Editor: Added video transcoding buttons to design menu

barklund opened the pull request:
## Context

This makes the buttons for mute and trim video functional.

Note that this also re-frames the mute icon inside its SVG file to allow it to be used for an icon button similarly to other icons. The result in the library looks the same as before.

## To-do

- [x] New mute icon ~?~
- [x] Tests

## Non-user-facing changes

| Video type | Screenshot |
|-|-|
| Local video with audio | ![Screen Shot 2022-03-03 at 12 38 47](https://user-images.githubusercontent.com/637548/156620716-4bcd3ba8-6bf5-4799-873a-348b0f9c9ac0.png) |
| Local video without audio | ![Screen Shot 2022-03-03 at 12 32 46](https://user-images.githubusercontent.com/637548/156620732-544640de-1dcf-4a55-b1e4-783c9d58c243.png) |
| Non-local video | ![Screen Shot 2022-03-03 at 12 32 50](https://user-images.githubusercontent.com/637548/156620749-827b77c7-6093-42df-8aa3-11f45f9a57b3.png) |

## Testing Instructions

<!-- ignore-task-list-start -->
- [x] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #10634

Commit:
Editor: Added video transcoding buttons to design menu

Commit:
Editor: Added video transcoding buttons to design menu

miina commented on packages/story-editor/src/components/floatingMenu/elements/mute.js:
```diff
@@ -24,13 +24,25 @@ import { __ } from '@googleforcreators/i18n';
-      title={__('mute video', 'web-stories')}
-      onClick={() => {}}
+      title={title}
+      isDisabled={isDisabled || isMuting}
```
```suggestion
      disabled={isDisabled || isMuting}
```

miina replied:
Hm, this is not related to this PR but in case of some videos, I get this console error:
<img width="1378" alt="Screenshot 2022-03-03 at 11 29 22" src="https://user-images.githubusercontent.com/3294597/156536352-3379268a-110c-4e41-a3fb-6df1a35bb31e.png">

Unfortunately I'm not able to repro it consistently. @spacedmonkey is this error something you've seen recently?

spacedmonkey replied:
I was not able to replicate that. Can provide steps to replicate or a video or something.

swissspidy replied:
@miina Perhaps the original unmuted video got deleted in the meantime?

miina replied:
@swissspidy Hm, I thought that issue got fixed somewhere 🤔 Can't find it currently though for muting, saw just this for trimming: https://github.com/GoogleForCreators/web-stories-wp/issues/10245

barklund replied:
Hmm, this seems unrelated to this PR. I'm not seeing any issues. I fixed the original comment here in 0c0d1a483a and will mark as resolved for now.

miina commented on packages/story-editor/src/components/floatingMenu/elements/trim.js:
```diff
@@ -24,13 +24,25 @@ import { __ } from '@googleforcreators/i18n';
-      title={__('Trim video', 'web-stories')}
-      onClick={() => {}}
+      title={title}
+      isDisabled={isDisabled || isTrimming}
```
```suggestion
      disabled={isDisabled || isTrimming}
```

barklund replied:
Fixed in 0c0d1a483a.

miina reviewed (approved):
LGTM (pending the better icon / tests) but would be great if @spacedmonkey could also review.

spacedmonkey commented on packages/story-editor/src/components/floatingMenu/elements/shared/useVideoTranscoding.js:
```diff
@@ -0,0 +1,77 @@
+  );
+
+  const { isMuted, id: resourceId = 0 } = resource;
+  const isTrimming = isResourceTrimming(resourceId);
```
@timarney Will this line need to be changed?

swissspidy replied:
Yes, the function is being renamed in that other PR

spacedmonkey replied:
And use the element id.

timarney replied:
yes
https://github.com/GoogleForCreators/web-stories-wp/pull/10712/files#diff-d581131a68dd353c626d1714ea3b4d65115979020ed22b61c7e5d4a0a8fed0ddR96

Commit:
Replaced muted icon

Commit:
Replaced muted icon

barklund commented on packages/design-system/src/icons/muted.svg:
```diff
@@ -1 +1,11 @@
-<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="#000000"><path d="M0 0h24v24H0V0z" fill="none"/><path d="M4.34 2.93L2.93 4.34 7.29 8.7 7 9H3v6h4l5 5v-6.59l4.18 4.18c-.65.49-1.38.88-2.18 1.11v2.06c1.34-.3 2.57-.92 3.61-1.75l2.05 2.05 1.41-1.41L4.34 2.93zM10 15.17L7.83 13H5v-2h2.83l.88-.88L10 11.41v3.76zM19 12c0 .82-.15 1.61-.41 2.34l1.53 1.53c.56-1.17.8 ...
\ No newline at end of file
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
```
This new icon is pretty big, bitwise. I did optimize it as much as possible using SVGO, but it's still annoyingly large. Could be handcrafted to be much smaller, but I'm not going to bother with it.

Commit:
Fixed disabling buttons while transcoding

Commit:
Fixed disabling buttons while transcoding

Commit:
Added unit tests for video options buttons

Karma tests are not possible because transcoding does not work in the headless browser.

Commit:
Added unit tests for video options buttons

Karma tests are not possible because transcoding does not work in the headless browser.

merapi reviewed (approved):
👌🏻

spacedmonkey reviewed (changes requested):
This PR needs to be tweaked now that https://github.com/GoogleForCreators/web-stories-wp/pull/10712 is merged.

Commit:
Merge branch 'main' into add/10634-floating-video

Commit:
Refactored video transcoding logic into custom hook

barklund commented on packages/story-editor/src/components/floatingMenu/elements/shared/useVideoTranscoding.js:
```diff
@@ -0,0 +1,77 @@
+  const { isTranscodingEnabled } = useFFmpeg();
+  const {
+    muteExistingVideo,
+    isResourceTrimming,
```
@timarney, @spacedmonkey: I'm getting [an error for this one in karma](https://github.com/GoogleForCreators/web-stories-wp/runs/5421645159?check_suite_focus=true#step:7:837)? Can this not be run in karma or? Can the function at least exist even if not doing the right thing?

swissspidy replied:
@barklund The function is now called `isElementTriming` and accepts an element ID.

spacedmonkey replied:
See https://github.com/GoogleForCreators/web-stories-wp/pull/10784/files#r818585757 and https://github.com/GoogleForCreators/web-stories-wp/pull/10712 for context.

spacedmonkey commented on packages/story-editor/src/components/floatingMenu/elements/shared/useVideoTranscoding.js:
```diff
@@ -0,0 +1,77 @@
+import useFFmpeg from '../../../../app/media/utils/useFFmpeg';
+import useProperties from './useProperties';
+
+function useVideoTranscoding() {
```
This hook could be very useful for [VideoOptionsPanel](https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/story-editor/src/components/panels/design/videoOptions/videoOptions.js). In fact is mostly a repeat of that code. Why not make this code more reusable and use it in both places? It will make future changes and refactors much easier.

barklund replied:
That's exactly what I had done and wanted you to review. And forgot to push 🤦 .

So, here it is (for real this time). Sorry for the inconvenience!

swissspidy replied:
Any thoughts on using a less generic name? At first glance `useVideoTranscoding` could easily be mistaken for `useFFmpeg` (in fact, that is its former name).

Maybe even `useVideoElementTranscoding` is an improvement, but I'm sure there are better names.

spacedmonkey commented on packages/story-editor/src/components/floatingMenu/elements/shared/useVideoTranscoding.js:
```diff
@@ -0,0 +1,77 @@
+    })
+  );
+
+  return {
```
I think this maybe should be broken up into actions and state. It will make clearer to know what is state and actions.

barklund replied:
I have now done that in 77bff9e with tests fixed in 62c72f9.

Commit:
Merge branch 'add/10634-floating-video' of github.com:GoogleForCreators/web-stories-wp into add/10634-floating-video

Commit:
Merge branch 'main' into add/10634-floating-video

Commit:
Refactored hook signature

Commit:
Fixed tests post refactor

spacedmonkey commented on packages/story-editor/src/components/panels/design/videoOptions/useVideoTranscoding.js:
```diff
@@ -0,0 +1,84 @@
+/*
```
Feels like this hook should exist [here](https://github.com/GoogleForCreators/web-stories-wp/tree/main/packages/story-editor/src/app/media/utils). It is used for multiple commpents and should not live with any component.

barklund replied:
The hook uses functionality from the story-editor package, so I didn't see an easy way to move to another package. We could extract that functionality, but that would make the hook less reusable (as the removed logic would have to be duplicated).

swissspidy replied:
Jonny pointed to `packages/story-editor/src/app/media/utils`, not another package. But not sure it's a great fit.

barklund replied:
Ah yes, I see now. Yeah, that could work. I have no strong opinions either way. It's only ever (according to current plans at least) going to be used in the two locations where it is now, so anything that works for those are ok with me.

barklund replied:
Moved in fc958e66c4.

spacedmonkey commented on packages/story-editor/src/components/panels/design/videoOptions/useVideoTranscoding.js:
```diff
@@ -0,0 +1,84 @@
+import useVideoTrim from '../../../videoTrim/useVideoTrim';
+import useFFmpeg from '../../../../app/media/utils/useFFmpeg';
+
+function useVideoTranscoding({ resource, elementId, isSingleElement = true }) {
```
Name of this hook feels a little generic here. Can we do better? @swissspidy any thoughts?

barklund replied:
I like @swissspidy's suggestion of `useVideoElementTranscoding`. I was struggling a bit with naming here, have no better suggestions.

`useTrimOrMute`? 🤷

swissspidy replied:
Let's use `useVideoElementTranscoding` for now then, even though it's a bit long

barklund replied:
Done in b1f31cba21.

Commit:
Renamed video transcoding hook

Commit:
Moved video transcoding hook
